lib/spack/spack/build_environment.py

```diff
@@ -323,6 +336,20 @@ def set_build_environment_variables(pkg, env, dirty):
     link_deps       = set(pkg.spec.traverse(root=False, deptype=('link')))
     build_link_deps = build_deps | link_deps
     rpath_deps      = get_rpath_deps(pkg)
+    build_run_deps = set()
+    for build_dep in build_deps:
+        build_run_deps.update(build_dep.traverse(deptype='run'))
+
+    # External packages may be installed in a prefix which contains many other
+    # package installs. To avoid having those installations override
+    # Spack-installed packages, they are placed at the end of search paths.
+    # System prefixes are removed entirely later on since they are already
+    # searched.
+    build_deps = _place_externals_last(build_deps)
+    link_deps = _place_externals_last(link_deps)
+    build_link_deps = _place_externals_last(build_link_deps)
+    rpath_deps = _place_externals_last(rpath_deps)
+    build_run_deps = _place_externals_last(build_run_deps)
```

Comment:
Not a request for change, just wanted to note that we are traversing lists 10 times (2 for each call). If this ever becomes a problem for large specs, we might need to refactor this series of call.